src/radical/pilot/agent/lm/srun.py

```diff
@@ -0,0 +1,133 @@
+    the task uses (a part of) a single node, or the task is using multiple nodes
+    uniformely.  Core pinning is only available on tasks which use exactly one
+    full node (and in that case becomes useless for our purposes).
+    
+    We use srun in the following way:
+
+        IF    task <= nodesize  
+        OR    task is uniformel
+        THEN  enforce node placement
+        ELSE  leave *all* placement to slurm
+    '''
+
+    # --------------------------------------------------------------------------
+    #
+    def __init__(self, cfg, session):
+
+        LaunchMethod.__init__(self, cfg, session)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def _configure(self):
+
+        self.launch_command = ru.which('srun')
+
+        out, err, ret = ru.sh_callout('%s -V' % self.launch_command)
+        if ret:
+            raise RuntimeError('cannot use srun [%s] [%s]' % (out, err))
+
+        self._version = out.split()[1]
+        self._log.debug('using srun from %s [%s]',
+                        self.launch_command, self._version)
+
+
+    # --------------------------------------------------------------------------
+    #
+    def construct_command(self, cu, launch_script_hop):
+
+        slots        = cu.get('slots')
+        cud          = cu['description']
+        task_exec    = cud['executable']
+        task_env     = cud.get('environment') or dict()
+        task_args    = cud.get('arguments')   or list()
+        task_argstr  = self._create_arg_string(task_args)
+        sbox         = ru.Url(cu['unit_sandbox']).path
+
+        # Construct the executable and arguments
+        if task_argstr: task_command = "%s %s" % (task_exec, task_argstr)
+        else          : task_command = task_exec
+
+        env_string = ''
+        env_list   = self.EXPORT_ENV_VARIABLES + task_env.keys()
+        if env_list:
+
+            env_string = '-export="%s"' % ','.join(env_list)
+
+
+        if not slots:
+            # leave placement to srun
+            nprocs    = cud['cpu_processes']
+            ncores    = cud['cpu_threads']
+            placement = ''
+
+
+        else:
+            
+            # Extract all the hosts from the slots
+            hostlist = list()
+            uniform  = True
+            chunk    = None
+            for node in slots['nodes']:
+
+                this_chunk = [len(node['core_map']),
+                              len(node['gpu_map' ])]
+
+                if not chunk:
+                    chunk = this_chunk
+
+                if chunk != this_chunk:
+                    uniform = False
+                    break
+
+                for _ in node['core_map']:
+                    hostlist.append(node['name'])
+
+                for _ in node['gpu_map']:
+                    hostlist.append(node['name'])
+
+            placement = ""
+            if uniform and len(hostlist) > 1:
```

Comment:
In the description, `srun` has the condition: `IF    task <= nodesize    OR    task is uniformel`. `task <= nodesize` is implicitly applied? because I don't find it.

Reply:
Yes, indeed - good catch.  This code does not attempt to place the task on a specific node in the single node case, but leaves the placement to srun (or rather slurm).  The same as for non-uniform multi-node tasks.  I  agree that this is inconsistent - I'll add the placement for single-node tasks (which are also uniform by definition, so that just means removing the check for hostlist length).  Thanks!